### PR TITLE
Extracted kore functionality to a proper backend.

### PR DESCRIPTION
**Pull request by traiansf**
Also added `krun` functionality.

### CODE REVIEW

**Comment by dwightguth** [comment]
I actually don't want the kore code to be in a separate backend... It should stay part of the kernel since it's now becoming part of the core of the new K.

That said, you should introduce a haskell backend which assumes that kore is installed on the path and invokes it from krun.

**Comment by traiansf** [comment]
@dwightguth please take another look.  I attempted to refactor it as you suggested.

**Comment by traiansf** [comment]
@dwightguth I went on and added more commits to this PR. it now also converts heating/cooling rules to equations and changes the encoding of owise rules as per the design decision.  The commits are pretty self-contained.

**Comment by traiansf** [comment]
@dwightguth I had a chat with Grigore on Thursday about how to encode subsorts and we got to the conclusion that the best way to do that is to add definedness (totality) axioms for `inj`, which is what I attempted to do in the commit above.

File: haskell-backend/src/main/java/org/kframework/backend/haskell/HaskellRewriter.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        this.files = files;
+        this.def = def;
+        this.haskellKRunOptions = haskellKRunOptions;
+        this.options = options;
+    }
+
+    @Override
+    public Rewriter apply(Module module) {
+        if (!module.equals(def.executionModule())) {
+            throw KEMException.criticalError("Invalid module specified for rewriting. Ocaml backend only supports rewriting over" +
```
Comment (dwightguth) [inline haskell-backend/src/main/java/org/kframework/backend/haskell/HaskellRewriter.java]: this should probably say Haskell, not Ocaml.

File: kernel/src/main/java/org/kframework/kast/KastFrontEnd.java [hunk 2]
```diff
@@ -114,7 +114,7 @@ public int run() {
                 parsed = new ExpandMacros(compiledMod, files, def.kompileOptions, false).expand(parsed);
             }
             if (options.kore) {
-              ModuleToKORE converter = new ModuleToKORE(compiledMod, files, def.topCellInitializer);
+              ModuleToKORE converter = new ModuleToKORE(compiledMod, files, null);
```
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/kast/KastFrontEnd.java]: why was this done?

**Review by dwightguth — APPROVED** [review]
looks good but I have one minor suggestion and one comment you should probably address before you merge.

**Review by dwightguth — CHANGES_REQUESTED** [review]
actually I take it back, something about your changes broke the llvm backend.

**Review by dwightguth — APPROVED** [review]

**Comment by dwightguth** [comment]
haha never mind, it was trivial to fix.

File: kernel/src/main/java/org/kframework/kast/KastFrontEnd.java (hunk 2 above)
Comment (traiansf) [reply kernel/src/main/java/org/kframework/kast/KastFrontEnd.java]: Sorry about that!  I think it happened while doing conflict resolution during the merge; for some reason I didn't noticed I can use `def.topCellInitializer` there.  I'm reverting the change.

**Comment by traiansf** [comment]
@dwightguth I fixed the issues you pointed out, but I think it's you who need to do the merge.

**Pull request merged**